### DISCUSSION TITLE
getIPAddr() for bind()

### DISCUSSION THREAD

**Discussion by hjpaul17** [ANLAB-KAIST/KENSv3#190] [Project Q&A 2. Transport Layer (TCP)] [answered]
I have some questions regarding bind().

1. If specific IP address has given for the argument, socket should bind to that IP as far as it has not overlayed. However, if the IP address of the argument is INADDR_ANY, what should I do? 
2. For checking validity of binding, you should check if the same IP address and port number are already in use. In #163 instructor advised for using getIPAddr(), but I don't get how should I use it. 
My understanding for getIPAddr() is that it receives a port number and returns an IP address that is using that port number. If so, it only returns one IP address that is using the port number, then we don't know if there is an another IP address using that port number. 
If using getIPAddr() is not the answer, then should I implement a data structure that stores all IP and port number that are already in use?
3. If we should use getIPAddr(), then should we also set using setIPAddr()?

**Comment by Pusnow** [comment, marked as answer]
> 1. If specific IP address has given for the argument, socket should bind to that IP as far as it has not overlayed. However, if the IP address of the argument is INADDR_ANY, what should I do?

You should bind it to all IP addresses. If there is already bound address/port, it would fail.

> 2. ...

There are two kinds of ports. See #178.

> should I implement a data structure that stores all IP and port number that are already in use?

Yes.

> 3. If we should use getIPAddr(), then should we also set using setIPAddr()?

No, you don't need to use `setIPAddr()`.
